### PR TITLE
Sync master with GPU

### LINKED ISSUE #435: Sync master with GPU
Add an optional CUDA flag to include corresponding libraries while building Vosk.

Python wrapper should be adjusted with missing GPU methods as well.

### PR DESCRIPTION
**Pull request by sskorol**
- Added an optional HAVE_CUDA flag to build Vosk with GPU support.
- Added missing GpuInit/Instantiate python wrappers.

Fixes #435

### CODE REVIEW

File: python/vosk/__init__.py [hunk 1]
```diff
@@ -69,3 +69,17 @@ def FinalResult(self):
...
+    if IsGpuEnabled():
+        _c.vosk_gpu_init()
+
+
+def GpuInstantiate():
+    if IsGpuEnabled():
+        _c.vosk_gpu_instantiate()
+
+
+def IsGpuEnabled():
```
Comment (nshmyrev) [inline python/vosk/__init__.py]: Лучше в С коде использовать заглушки 

```
void vosk_gpu_init() {
#if HAVE_CUDA
...
#endif
}
```

а сами функции оставить пустыми. Кстати, не вижу vosk_api.h часть тут в изменениях, забылась?

File: src/Makefile [hunk 2]
```diff
@@ -35,6 +38,20 @@ LIBS= \
 	$(OPENFST_ROOT)/lib/libfst.a \
 	$(OPENFST_ROOT)/lib/libfstngram.a
 
+# Since CUDA 10.1 some libs are located in /usr/lib/[ARCH]. So you should symlink them with $(CUDA_ROOT)/lib64 or adjust ld.so.conf.
+ifeq ($(HAVE_CUDA), 1)
+    LIBS:=$(LIBS) \
```
Comment (nshmyrev) [inline src/Makefile]: Можно += использовать

File: src/Makefile [hunk 3]
```diff
@@ -35,6 +38,20 @@ LIBS= \
 	$(OPENFST_ROOT)/lib/libfst.a \
 	$(OPENFST_ROOT)/lib/libfstngram.a
 
+# Since CUDA 10.1 some libs are located in /usr/lib/[ARCH]. So you should symlink them with $(CUDA_ROOT)/lib64 or adjust ld.so.conf.
```
Comment (nshmyrev) [inline src/Makefile]: В kaldi.mk нет /usr/lib/arch, и так всё работает. Должно работать и тут.

File: src/Makefile [hunk 4]
```diff
@@ -10,7 +13,7 @@ VOSK_SOURCES= \
     spk_model.cc \
     vosk_api.cc
 
-CFLAGS=-g -O2 -std=c++17 -fPIC -DFST_NO_DYNAMIC_LINKING -I. -I$(KALDI_ROOT)/src -I$(OPENFST_ROOT)/include -I$(OPENBLAS_ROOT)/include
+CFLAGS=-g -O2 -std=c++17 -fPIC -DFST_NO_DYNAMIC_LINKING -DHAVE_CUDA=$(HAVE_CUDA) -I. -I$(KALDI_ROOT)/src -I$(OPENFST_ROOT)/include -I$(OPENBLAS_ROOT)/include
```
Comment (nshmyrev) [inline src/Makefile]: Эту часть можно оставить внизу в if HAVE_CUDA, не менять по умолчанию.

File: python/vosk/__init__.py (hunk 1 above)
Comment (sskorol) [reply python/vosk/__init__.py]: Да, забыл перенести из GPU ветки. Исправил.

File: src/Makefile (hunk 2 above)
Comment (sskorol) [reply src/Makefile]: Исправил.

File: src/Makefile (hunk 3 above)
Comment (sskorol) [reply src/Makefile]: Убрал.

File: src/Makefile (hunk 4 above)
Comment (sskorol) [reply src/Makefile]: Перенес.

File: src/vosk_api.h [hunk 5]
```diff
@@ -38,6 +38,7 @@ typedef struct VoskSpkModel VoskSpkModel;
  *  which represent decoded information - words, confidences, times, n-best lists,
  *  speaker information and so on */
 typedef struct VoskRecognizer VoskRecognizer;
+typedef struct VoskGpu VoskGpu;
```
Comment (nshmyrev) [inline src/vosk_api.h]: Not needed anymore right?

File: src/vosk_api.h [hunk 6]
```diff
@@ -212,6 +213,9 @@ void vosk_recognizer_free(VoskRecognizer *recognizer);
  */
 void vosk_set_log_level(int log_level);
 
+void vosk_gpu_init();
```
Comment (nshmyrev) [inline src/vosk_api.h:220]: Could you add doxygen comments like other methods please?

File: src/vosk_api.cc [hunk 7]
```diff
@@ -100,3 +100,22 @@ void vosk_set_log_level(int log_level)
...
     SetVerboseLevel(log_level);
 }
+
+#if HAVE_CUDA
+#include "cudamatrix/cu-device.h"
+#endif
+
+void vosk_gpu_init()
+{
+    #if HAVE_CUDA
```
Comment (nshmyrev) [inline src/vosk_api.cc]: preprocessor directives usually start from beginning of the line without indent.

File: src/vosk_api.h (hunk 5 above)
Comment (sskorol) [reply src/vosk_api.h]: I don't know. It was added on May 2, 2020. And still lives in the GPU branch. So I have no idea if it's something intended to be used in the future, or just a dead code that is not required anymore. But if you say it's redundant, I'll remove it of course.

File: src/vosk_api.h (hunk 6 above)
Comment (sskorol) [reply src/vosk_api.h:220]: Will add. But there's a quick question.

According to [Kaldi sources](https://github.com/kaldi-asr/kaldi/blob/master/src/cudamatrix/cu-device.h#L82-L91), `CuDevice::Instantiate()` returns an instance of `CuDevice` for a current thread. I understand why do we need to call it within `vosk_gpu_init()`, as it selects gpu id and allows multithreading.

But what I can't get is why do we need a separate `vosk_gpu_instantiate()` method, which just calls a `CuDevice` initializer, but nothing more? Isn't it enough just to call `vosk_gpu_init()` which already encapsulates `CuDevice::Instantiate()`? These 2 methods seem very confusing at the moment.

File: src/vosk_api.cc (hunk 7 above)
Comment (sskorol) [reply src/vosk_api.cc]: Will fix.

File: src/vosk_api.h (hunk 5 above)
Comment (sskorol) [reply src/vosk_api.h]: Removed.

File: src/vosk_api.cc (hunk 7 above)
Comment (sskorol) [reply src/vosk_api.cc]: Reformatted.

File: src/vosk_api.h (hunk 6 above)
Comment (sskorol) [reply src/vosk_api.h:220]: Added docs + renamed `vosk_gpu_instantiate`.

**Pull request merged**